// testcafe/lib/native-automation/request-pipeline/test-run-bridge.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const connection_1 = __importDefault(require("../../browser/connection"));
class TestRunBridge {
    constructor(browserId) {
        this._browserId = browserId;
    }
    getBrowserConnection() {
        return connection_1.default.getById(this._browserId);
    }
    getCurrentTestRun() {
        const browserConnection = this.getBrowserConnection();
        return browserConnection.getCurrentTestRun();
    }
    getSessionId() {
        var _a;
        return ((_a = this.getCurrentTestRun()) === null || _a === void 0 ? void 0 : _a.id) || '';
    }
    getUserScripts() {
        const currentTestRun = this.getCurrentTestRun();
        if (!currentTestRun)
            return [];
        return currentTestRun.injectable.userScripts;
    }
    getInjectableScripts() {
        const currentTestRun = this.getCurrentTestRun();
        if (!currentTestRun)
            return [];
        return currentTestRun.injectable.scripts;
    }
    getInjectableStyles() {
        const currentTestRun = this.getCurrentTestRun();
        if (!currentTestRun)
            return [];
        return currentTestRun.injectable.styles;
    }
    async getTaskScript({ isIframe }) {
        const browserConnection = this.getBrowserConnection();
        const proxy = browserConnection.browserConnectionGateway.proxy;
        const windowId = browserConnection.activeWindowId;
        // @ts-ignore
        return await this.getCurrentTestRun().session.getTaskScript({
            referer: '',
            cookieUrl: '',
            withPayload: true,
            serverInfo: proxy.server1Info,
            windowId,
            isIframe,
        });
    }
}
exports.default = TestRunBridge;
module.exports = exports.default;
