// testcafe/lib/client/core/utils/array.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCommonElement = exports.equals = exports.remove = exports.find = exports.from = exports.isArray = exports.every = exports.join = exports.concat = exports.reduce = exports.reverse = exports.some = exports.indexOf = exports.forEach = exports.unshift = exports.splice = exports.slice = exports.map = exports.filter = void 0;
const hammerhead_1 = require("../deps/hammerhead");
const ARRAY_METHODS_PREFIX = 'array';
function createNativeMethodWrapper(methodName) {
    const nativeMethodName = ARRAY_METHODS_PREFIX + methodName.charAt(0).toUpperCase() + methodName.slice(1);
    const nativeMethod = hammerhead_1.nativeMethods[nativeMethodName];
    return (...args) => nativeMethod.call(...args);
}
exports.filter = createNativeMethodWrapper('filter');
exports.map = createNativeMethodWrapper('map');
exports.slice = createNativeMethodWrapper('slice');
exports.splice = createNativeMethodWrapper('splice');
exports.unshift = createNativeMethodWrapper('unshift');
exports.forEach = createNativeMethodWrapper('forEach');
exports.indexOf = createNativeMethodWrapper('indexOf');
exports.some = createNativeMethodWrapper('some');
exports.reverse = createNativeMethodWrapper('reverse');
exports.reduce = createNativeMethodWrapper('reduce');
exports.concat = createNativeMethodWrapper('concat');
exports.join = createNativeMethodWrapper('join');
exports.every = createNativeMethodWrapper('every');
function isArray(arg) {
    return hammerhead_1.nativeMethods.objectToString.call(arg) === '[object Array]';
}
exports.isArray = isArray;
function from(arg, ...args) {
    if (hammerhead_1.nativeMethods.arrayFrom)
        return hammerhead_1.nativeMethods.arrayFrom(arg, ...args);
    // NOTE: this logic is for IE
    const arr = [];
    const length = arg.length;
    for (let i = 0; i < length; i++)
        arr.push(arg[i]);
    return arr;
}
exports.from = from;
function find(arr, callback) {
    if (hammerhead_1.nativeMethods.arrayFind)
        return hammerhead_1.nativeMethods.arrayFind.call(arr, callback);
    // NOTE: this logic is for IE
    const length = arr.length;
    for (let i = 0; i < length; i++) {
        if (callback(arr[i], i, arr))
            return arr[i];
    }
    return null;
}
exports.find = find;
function remove(arr, item) {
    const index = hammerhead_1.nativeMethods.arrayIndexOf.call(arr, item);
    if (index > -1)
        hammerhead_1.nativeMethods.arraySplice.call(arr, index, 1);
}
exports.remove = remove;
function equals(arr1, arr2) {
    if (arr1.length !== arr2.length)
        return false;
    for (let i = 0, l = arr1.length; i < l; i++) {
        if (arr1[i] !== arr2[i])
            return false;
    }
    return true;
}
exports.equals = equals;
function getCommonElement(arr1, arr2) {
    for (let i = 0; i < arr1.length; i++) {
        for (let t = 0; t < arr2.length; t++) {
            if (arr1[i] === arr2[t])
                return arr1[i];
        }
    }
    return null;
}
exports.getCommonElement = getCommonElement;
