// testcafe/lib/native-automation/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const request_pipeline_1 = __importDefault(require("./request-pipeline"));
const add_custom_debug_formatters_1 = __importDefault(require("./add-custom-debug-formatters"));
const debug_loggers_1 = require("../utils/debug-loggers");
const session_storage_1 = __importDefault(require("./session-storage"));
class NativeAutomation {
    constructor(browserId, client, options) {
        this._client = client;
        this.options = options;
        this.requestPipeline = new request_pipeline_1.default(browserId, client, options);
        this.sessionStorage = new session_storage_1.default(browserId, client, options);
        (0, add_custom_debug_formatters_1.default)();
    }
    _onContextStorageSyncHandler({ sessionStorage, testRunId, frameDriverId }) {
        if (sessionStorage) {
            this.requestPipeline.contextStorage = this.requestPipeline.contextStorage || {};
            this.requestPipeline.contextStorage[testRunId] = this.requestPipeline.contextStorage[testRunId] || {};
            this.requestPipeline.contextStorage[testRunId][frameDriverId] = sessionStorage;
        }
    }
    _onContextStorageTestRunDoneHandler({ testRunId }) {
        if (this.requestPipeline.contextStorage)
            delete this.requestPipeline.contextStorage[testRunId];
    }
    _addEventListeners() {
        this.sessionStorage.on('contextStorageSync', this._onContextStorageSyncHandler.bind(this));
        this.sessionStorage.on('contextStorageTestRunDone', this._onContextStorageTestRunDoneHandler.bind(this));
    }
    async start() {
        (0, debug_loggers_1.nativeAutomationLogger)('starting');
        for (const apiSystem of this.apiSystems)
            await apiSystem.start();
        this._addEventListeners();
        (0, debug_loggers_1.nativeAutomationLogger)('started');
    }
    async dispose() {
        this.requestPipeline.stop();
        await this.requestPipeline.dispose();
        (0, debug_loggers_1.nativeAutomationLogger)('nativeAutomation disposed');
    }
    get apiSystems() {
        return [
            this.requestPipeline,
            this.sessionStorage,
        ];
    }
}
exports.default = NativeAutomation;
module.exports = exports.default;
