// mongo-object/dist/cjs/util.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const expect_1 = __importDefault(require("expect"));
const util_js_1 = require("./util.js");
const expect = expect_1.default.default;
describe('util', () => {
    it('cleanNulls', () => {
        const date = new Date();
        const cleaned = (0, util_js_1.cleanNulls)({
            a: undefined,
            b: undefined,
            c: null,
            d: '',
            e: 'keep me',
            f: {
                a: undefined,
                b: undefined,
                c: null,
                d: '',
                e: 'keep me'
            },
            g: {
                a: null
            },
            h: {
                a: date
            }
        });
        expect(cleaned).toEqual({
            e: 'keep me',
            f: { e: 'keep me' },
            h: { a: date }
        });
    });
    it('cleanNulls with arrays', () => {
        const cleaned = (0, util_js_1.cleanNulls)({
            a: {
                b: [
                    {
                        c: null,
                        d: '',
                        e: undefined
                    },
                    {
                        c: null,
                        d: '',
                        e: [null, 'keep', '', undefined]
                    }
                ]
            }
        });
        expect(cleaned).toEqual({
            a: {
                b: [
                    undefined,
                    {
                        e: [undefined, 'keep']
                    }
                ]
            }
        });
    });
    it('reportNulls', () => {
        const report = (0, util_js_1.reportNulls)({
            a: undefined,
            b: undefined,
            c: null,
            d: '',
            e: 'keep me'
        });
        expect(report).toEqual({
            a: '',
            b: '',
            c: '',
            d: ''
        });
    });
});
